### ISSUE TITLE
Class Imbalance in Open Images Dataset

### ISSUE THREAD

**Issue by sophia-wright-blue** [qfgaohao/pytorch-ssd#13]
the Open Images Dataset has severe class imbalance - the label Person has 800,000 images, while some other images have 500 images. If i select 10 labels with different no. of images in each label, what would be the best way to address this imbalance?

thanks,

**Comment by qfgaohao** [comment]
Hi @sophia-wright-blue , I encountered the exactly same problem. There are two ways to deal with it.

1. Add '--balance_data' when you run train_ssd.py. The program will automatic down-sample more frequent examples.

2. Down-sample when  prepare training data. Run sth like the follows:

python open_images_downloader.py --root ~/data/open_images --class_names "Apple,Banana,Grape:group,**Human face::0.01**" --num_workers 20

The above command will only download 1 percent of Human face examples. The format of class name can be "className[:group:[percentage]]" . "group" means only downloading examples with the attribute 'IsGroupOf', '~group' means excluding examples with the attribute 'IsGroupOf'. Empty means both. Percentage means the percentage of images you want to download.

I am still improving this script and its manual. I will update the readme file when I see it is properly done.

**Comment by sophia-wright-blue** [comment]
thanks for replying @qfgaohao , i see in your readme that you mention ``` --balance_data ```, i should've seen that;

thanks for the info about downloading a certain percentage of the data - that is very helpful,

thanks again for releasing this repo - it is helping me learn a lot;

i have a related follow-up that you may be able to shed some light on: i'm trying to understand the speed-accuracy tradeoffs of real-time object detection models. i have tested YOLO-v3 vs SSD and i'm convinced about the accuracy advantage of SSD;

i'm now exploring SSD vs RetinaNet; in the RetinaNet paper, they mention that RetinaNet has better speed-accuracy tradeoff than SSD, but in my tests on RetinaNet (i'm using this repo - https://github.com/fizyr/keras-retinanet), it takes 0.5-2.0 seconds to run inference on an image, which is too slow real-time object detection; (i'm using a Tesla K80 GPU)

do you have experience with the speed-accuracy results on RetinaNet? i'd greatly appreciate any guidance on this,

thanks,

**Comment by qfgaohao** [comment]
Hi @sophia-wright-blue , the speed and accuracy of a model largely depends on the size of the base net it uses. Try to use a smaller base net such as MobileNet, if you want to increase the speed.

**Comment by sophia-wright-blue** [comment]
thanks for replying @qfgaohao , i'll try experimenting with a smaller base net, thanks!
